### ISSUE TITLE
support for environment variables

### ISSUE THREAD

**Issue by kmala** [kubernetes/helm#944]
Currently helm doesn't support environment variables because templating of the charts happens on the server side.Again people might like to pass the values through env than pass them as  values file. So, during a `helm install` we can grab the current env and pass it on to the server so that it can use while templating the charts .The env can be thought of as something similar to global values in the values.yaml.

**Comment by technosophos** [comment]
**tl;dr**: There are five options for how we can approach the environment variables issue. They are:

1. Send the user's environment to the server, re-enable the `env` command, but for those variables.
2. Add `--val NAME=VALUE` method of setting values on Helm
3. Expand environment variables on the file passed in using `-f` on `helm install -f values.yaml`.
4. Accept values.yaml on STDIN
5. Relegate this to external tools

Each of these has strengths and weaknesses. I personally think 4 or 5 are the only ones that maintain our current design goals.
----

Okay, now brace yourselves for a long one.

This issue is proving to be a little bit contentions. I removed environment variable support because it conflicted with two design goals of Helm:

1. Reproducible builds: Given a chart and values, a build should be reproducible (except for intentionally variable data, like timestamps)
2. Sandboxed chart interpretation: During chart rendering, a chart should only have access to data that was injected into the chart.

That was my original motivation.

There are currently five proposals for adding environment variable support back in. They are:

1. Send the user's environment to the server, re-enable the `env` command, but for those variables.
2. Add `--val NAME=VALUE` method of setting values on Helm
3. Expand environment variables on the file passed in using `-f` on `helm install -f values.yaml`.
4. Accept values.yaml on STDIN
5. Relegate this to external tools

Here are the strengths and weaknesses of each.

## Option 1: Send the environment to Tiller

The main advantage of this method is that it preserves (from the chart dev perspective) the templating capabilities from Helm Classic. One can simply do `{{env FOO}}` to get the value of an environment variable.

There are a few drawbacks. One is that the code for doing this in isolation (maintaining the sandbox) will be a little stilted. The second is that this will raise some very potent security concerns.  For example, I just looked at my environment variables, and there are several that I definitely would not want to be sent to Kubernetes and stored there. The third issue is that this seems to violate the reproducibility goal. Two developers can deploy the same chart with the same values file, and get (possibly unpredictably) different results.

## Option 2: Add command-line flags for passing values

This is one of the simplest to implement, and plays well with the UNIX paradigm. And it works like this: If I want to pass in extra values to Helm, I can do so on the CLI:

```
$ helm install -f values.yaml --val foo=$BAR mychart
```

Now, because the shell does interpolation for us, we automatically get environment variable support. All of the processing is done locally before the values are sent to the server, and existing calls like `helm get values` will show the computed values.

But, as @michelleN pointed out to me, this is a really cumbersome way of doing things. You'll end up with really unfriendly command lines like:

```
$ helm install -f values.yaml --val token=$MY_APP_TOKEN --val key=$MY_APP_SECRET_KEY mychart
```

Futhermore, it's not clear how we would support overriding namespaced and global params this way.

## Option 3: Interpolate `-f values.yaml`

The third option is to do environment variable interpolation over the file passed in with the `-f` flag. So say we had `helm install -f foo.yaml bar`, `foo.yaml` could contain lines like this:

```yaml
token: "$MY_APP_TOKEN"
key: "$MY_APP_SECRET_KEY"
```

Effectively, in the Go code, we'd run the above through `os.ExpandEnv()`. And this evaluation would be client-side, so the `foo.yaml` sent to the server would already have the results of the environment variable expansion. Thus `helm get` and other commands would work fine.

Just as with the previous solution, this only partially satisfies our design goal for reproducibility. And for that reason, I don't really like it. But... it's cleaner than the other options.

The other big drawback to this method is that it introduces another (weak) templating language, this time on the values file. And this would only work for files passed in with `-f` (because that's the only way to maintain reproducibility and maintain chart immutability).

## Option 4: Accept values files on STDIN

This method only adds the ability to read YAML values off of STDIN. That makes it easy to pipe in files from other sources which might perform other operations:

```
gen-values.sh | helm install nginx
```

This adds minor complexity to the CLI, but is a strongly UNIX-based way of doing things. I don't really know how many developers would use that sort of feature.

## Option 5: Do nothing, and encourage developers to solve externally

This model simply sticks with the assertion that the only way to pass configuration parameters into a release is using `-f FILE`. Developers who want to use environment variables would need to run some sort of pre-processing step that generates FILE.

**Comment by prydonius** [comment]
Thought I'd add my two cents on this one:

> Option 1: Send the environment to Tiller

I think this is dangerous for the security concerns you mentioned, especially since it won't be immediately obvious that this is happening. For reproducibility it's the worst case, without looking at the chart's template, it's not clear what env vars are actually being used. For these reasons, I strongly think we should avoid this option.

> Option 2: Add command-line flags for passing values

I like this, because it's similar to to what I was looking for in https://github.com/helm/helm-classic/issues/428. The downsides are as you mentioned, if you need to pass in a lot of env vars, it gets tedious, and also it's more difficult to share with other people. I do like that you are able to be more explicit with what env vars you're using, and also we're getting the shell to do all the work for us. It may be worth implementing this regardless of this use-case.

> Option 3: Interpolate -f values.yaml

This is an interesting idea, and eases some of my concerns about Option 1 in terms of explicitness. I agree with the concern about having another templating language, but if we keep it restricted to just expanding env vars as if the shell would, it might be ok.

> Option 4: Accept values files on STDIN

I'm not sure if there is much value in this approach, but I also don't think it would be that bad to support in general. Maybe it would read from STDIN if you provided something like `helm install . --values -`

> Option 5: Do nothing, and encourage developers to solve externally

This might be the way to go, if we want to be strict about the reproducibility goal, since all other options break that use-case.


I'm curious if a combination of 2 and 3 would be a good solution. I think if we end up going for option 5, option 2 could also work, since it's not built in the tool itself (the shell does the parsing of the env vars)

**Comment by kmala** [comment]
The reason for creating this issue is to simplify the process of passing the values for templating as passing through file is not as simple as setting the env variable in a deployment/installation scripts which rules out option 3,4 and 5.
I think option 2 will be  a good solution as it might be cumbersome for manual install but might make sense for people who have scripts.
i think this should take precedence over other ways of setting(global and values.yaml) or we can expand these values as env variables to the charts.

**Comment by technosophos** [comment]
Okay, I'll look at what it would take to implement Option 2.

**Comment by technosophos** [comment]
The pull request in #982 implements Option 2 by adding `helm install --set key1=val1,key2=val2`. And it supports a dotted notation for doing subchart: `helm install --set mysql.password=secret`.

**Comment by splashx** [comment]
Does anyone know if Option 3 could ever be implemented? (if that's welcomed) Having to run a re-templating engine to populate `values` in file is quite an overhead 😢

**Comment by ivanboiko** [comment]
I also vote for Option 3.

We have multiple environments (Test, Prod etc) and pass lots of env-specific values (db connections strings, external endpoints etc).
We don't want to hard-code all those values in the `values_<ENV>.yaml` files or anywhere in the git repository for that matter.
As those values live separate lifecycle, so we keep them in the CI tool and pass to the deployment job as shell env vars.

We were using `helm install --set ...` to pass them, but this approach is not very scalable.
As it's a mixture of process and app-specifics.
We have multiple micro-services and trying to use the same pipeline template in them.
Option 3 would be just perfect for us.

**Comment by MaRuifeng** [comment]
#982 by implementing option 2 solves the problem partially as it's not secure to pass in secret environment variables in that way. Below command would leak the secret information in the output of the `ps` command. It would be great to have either option 3 or option 4 such that secret information can be passed in from a different pipe. 

```
helm install --set password=${my_secret} ./mychart
```

Secret management is well requested in #2196 with great suggestions yet to be implemented.

**Comment by lorenzleutgeb** [comment]
With the release of Helm 3 to be expected soon, I'd like to ask how this changes the perspective on this question. Previously, in Helm 2, (as stated in the initial comment) templates were evaluated on the server-side (by Tiller). Since in Helm 3 everything is done client-side, it might be less of a problem to allow accessing the environment in templates (e.g. allow `env` for Go-style templates).

I saw that the documentation regarding writing templates changed quite a bit. One difference being that for Helm 2, the section [Chart Development Tips and Tricks: Know Your Template Functions](https://helm.sh/docs/charts_tips_and_tricks/#know-your-template-functions) talks about `env`, but for Helm 3, the section [Template Functions and Pipelines](https://v3.helm.sh/docs/topics/chart_template_guide/functions_and_pipelines/) does not mention `env`.

**Comment by bacongobbler** [comment]
See https://github.com/helm/helm/pull/5815

**Comment by millbj92** [comment]
@technosophos  I still really like the idea of pipes. You could ideally get the best of both worlds... Some small modifications to a lib like GoViper, and it could even be packaged with the helm binary as a parsing tool. I don't think devs would mind - and it would benefit to be able to have validation before the values enter the helm runtime. Even with the tools that we have in 2021; I would still love to see a feature like this.
